# UNIX Assignment## Project Background### Data Inspection* Become familiar with the files and to describe their structure and their dimensions (file size, number of columns, number of lines, etc...)*  The files are: fang_et_al_genotypes.txt: a published SNP data set including maize, teosinte (i.e., wild maize),and Tripsacum (a close outgroup to the genus Zea) individuals2. snp_position.txt: an additional data file that includes the SNP id (first column), chromosome location (third column), nucleotide location (fourth column) and other information for the SNPs genotyped in the fang_et_al_genotypes.txt file### Data ProcessingFor maize (Group = ZMMIL, ZMMLR, and ZMMMR in the third column of thefang_et_al_genotypes.txt file) we want 20 files in total:* 10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?* 10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -* 1 file with all SNPs with unknown positions in the genome (these need not be ordered in any particular way)* 1 file with all SNPs with multiple positions in the genome (these need not be ordered in any particular way)For teosinte (Group = ZMPBA, ZMPIL, and ZMPJA in the third column of the fang_et_al_genotypes.txt file) we want 20 files in total:* 10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?* 10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -* 1 file with all SNPs with unknown positions in the genome (these need not be ordered in any particular way)* 1 file with all SNPs with multiple positions in the genome (these need not be ordered in any particular way)## Project Script and Data### Data Inspection**fang et al genotypes**##  	$ wc fang_et_al_genotypes.txtNumber of Lines: 2783Number of Words: 2744038Number of Characters: 11051939 ##	$ awk '{print NF}' fang_et_al_genotypes.txt | sort -nu | tail -n 1Number of Columns: 986##	$ du -hs fang_et_al_genotypes.txtTotal File Size: 11M**snp position**##	$ wc snp_position.txtNumber of Lines: 984Number of Words: 13198Number of Characters: 83747##	$ awk '{print NF}' snp_position.txt | sort -nu | tail -n 1Number of Columns: 30##	$ du -hs snp_position.txtTotal File Size: 84K### Data Processing**Move Maize genotypes to separate file**## 	$ grep -E "(ZMMIL|ZMMLR|ZMMMR)" fang_et_al_genotypes.txt > maize_genotypes.txt**Inspect file**##	$ wc -l maize_genotypes.txt1573 lines##	$ awk '{print NF}' maize_genotypes.txt | sort -nu | tail -n 1986 columns**Move Teosinte genotypes to separate file**##	$ grep -E "(ZMPBA|ZMPIL|ZMPJA)" fang_et_al_genotypes.txt > teosinte_genotypes.txt**Inspect file**##	$ wc -l teosinte_genotypes.txt975 lines##	$ awk '{print NF}' teosinte_genotypes.txt | sort -nu | tail -n 1986 columns**Create header file and add to genotype files**##	$ head -n1 fang_et_al_genotypes.txt > headers.txt##	$ cat headers.txt maize_genotypes.txt > header_maize_genotypes.txt	$ cat headers.txt teosinte_genotypes.txt > header_teosinte_genotypes.txt**Transpose header genotype files prior to joining**##	$ awk -f transpose.awk header_maize_genotypes.txt > transposed_maize_genotypes.txt	$ awk -f transpose.awk header_teosinte_genotypes.txt > transposed_teosinte_genotypes.txt**Sort maize/teosinte transposed files**##	$ sort -k1,1 transposed_teosinte_genotypes.txt > sorted_teosinte.txt	$ sort -k1,1 transposed_maize_genotypes.txt > sorted_maize.txt**Separate out snp_position.txt for just the columns needed (1, 3, 4)**##	$ awk '{print $ "1\t" $3 "\t" $ "4\t"}' snp_position.txt > sorted_snp_position.txt**Sort file by column 1**##	$ sort -k1,1 sorted_snp_position.txt > sorted_snps.txt**Join the maize/teosinte file with the sorted snp file**##	$ join -t $'\t' -1 1 -2 1 sorted_snps.txt sorted_maize.txt > joined_file.txt	$ join -t $'\t' -1 1 -2 1 sorted_snps.txt sorted_teosinte.txt > joined_teosinte.txt**Separate out to 10 chromosomes, sort into increasing/decreasing order**##	$ for i in {1..10}; do awk '$2 == '$i'' joined_teosinte.txt > chr"$i"_teosinte_data.txt; done			$ for i in {1..10}; do awk '$2 == '$i'' joined_file.txt > chr"$i"_maize_data.txt; done		**Sort files into increasing order**##	$ sort -k3,3n chr1_maize_data.txt > chr1_maize_increase.txt	$ sort -k3,3n chr2_maize_data.txt > chr2_maize_increase.txt	$ sort -k3,3n chr3_maize_data.txt > chr3_maize_increase.txt	$ sort -k3,3n chr4_maize_data.txt > chr4_maize_increase.txt	$ sort -k3,3n chr5_maize_data.txt > chr5_maize_increase.txt	$ sort -k3,3n chr6_maize_data.txt > chr6_maize_increase.txt	$ sort -k3,3n chr7_maize_data.txt > chr7_maize_increase.txt	$ sort -k3,3n chr8_maize_data.txt > chr8_maize_increase.txt	$ sort -k3,3n chr9_maize_data.txt > chr9_maize_increase.txt	$ sort -k3,3n chr10_maize_data.txt > chr10_maize_increase.txt	$ sort -k3,3n chr1_teosinte_data.txt > chr1_teosinte_increase.txt	$ sort -k3,3n chr2_teosinte_data.txt > chr2_teosinte_increase.txt	$ sort -k3,3n chr3_teosinte_data.txt > chr3_teosinte_increase.txt	$ sort -k3,3n chr4_teosinte_data.txt > chr4_teosinte_increase.txt	$ sort -k3,3n chr5_teosinte_data.txt > chr5_teosinte_increase.txt	$ sort -k3,3n chr6_teosinte_data.txt > chr6_teosinte_increase.txt	$ sort -k3,3n chr7_teosinte_data.txt > chr7_teosinte_increase.txt	$ sort -k3,3n chr8_teosinte_data.txt > chr8_teosinte_increase.txt	$ sort -k3,3n chr9_teosinte_data.txt > chr9_teosinte_increase.txt	$ sort -k3,3n chr10_teosinte_data.txt > chr10_teosinte_increase.txt**Decreasing order:**##	$ sort -k3,3nr chr1_maize_data.txt > chr1_maize_decrease.txt	$ sort -k3,3nr chr2_maize_data.txt > chr2_maize_decrease.txt	$ sort -k3,3nr chr3_maize_data.txt > chr3_maize_decrease.txt	$ sort -k3,3nr chr4_maize_data.txt > chr4_maize_decrease.txt	$ sort -k3,3nr chr5_maize_data.txt > chr5_maize_decrease.txt	$ sort -k3,3nr chr6_maize_data.txt > chr6_maize_decrease.txt	$ sort -k3,3nr chr7_maize_data.txt > chr7_maize_decrease.txt	$ sort -k3,3nr chr8_maize_data.txt > chr8_maize_decrease.txt	$ sort -k3,3nr chr9_maize_data.txt > chr9_maize_decrease.txt	$ sort -k3,3nr chr10_maize_data.txt > chr10_maize_decrease.txt	$ sort -k3,3nr chr1_teosinte_data.txt > chr1_teosinte_decrease.txt	$ sort -k3,3nr chr2_teosinte_data.txt > chr2_teosinte_decrease.txt	$ sort -k3,3nr chr3_teosinte_data.txt > chr3_teosinte_decrease.txt	$ sort -k3,3nr chr4_teosinte_data.txt > chr4_teosinte_decrease.txt	$ sort -k3,3nr chr5_teosinte_data.txt > chr5_teosinte_decrease.txt	$ sort -k3,3nr chr6_teosinte_data.txt > chr6_teosinte_decrease.txt	$ sort -k3,3nr chr7_teosinte_data.txt > chr7_teosinte_decrease.txt	$ sort -k3,3nr chr8_teosinte_data.txt > chr8_teosinte_decrease.txt	$ sort -k3,3nr chr9_teosinte_data.txt > chr9_teosinte_decrease.txt	$ sort -k3,3nr chr10_teosinte_data.txt > chr10_teosinte_decrease.txt	**Replace unknowns with "-":**##	$ sed 's/?/-/g' chr1_maize_decrease.txt > chr1_maize_decrease_replace.txt	$ sed 's/?/-/g' chr2_maize_decrease.txt > chr2_maize_decrease_replace.txt	$ sed 's/?/-/g' chr3_maize_decrease.txt > chr3_maize_decrease_replace.txt	$ sed 's/?/-/g' chr4_maize_decrease.txt > chr4_maize_decrease_replace.txt	$ sed 's/?/-/g' chr5_maize_decrease.txt > chr5_maize_decrease_replace.txt	$ sed 's/?/-/g' chr6_maize_decrease.txt > chr6_maize_decrease_replace.txt	$ sed 's/?/-/g' chr7_maize_decrease.txt > chr7_maize_decrease_replace.txt	$ sed 's/?/-/g' chr8_maize_decrease.txt > chr8_maize_decrease_replace.txt	$ sed 's/?/-/g' chr9_maize_decrease.txt > chr9_maize_decrease_replace.txt	$ sed 's/?/-/g' chr10_maize_decrease.txt > chr10_maize_decrease_replace.txt	$ sed 's/?/-/g' chr1_teosinte_decrease.txt > chr1_teosinte_decrease_replace.txt	$ sed 's/?/-/g' chr2_teosinte_decrease.txt > chr2_teosinte_decrease_replace.txt	$ sed 's/?/-/g' chr3_teosinte_decrease.txt > chr3_teosinte_decrease_replace.txt	$ sed 's/?/-/g' chr4_teosinte_decrease.txt > chr4_teosinte_decrease_replace.txt	$ sed 's/?/-/g' chr5_teosinte_decrease.txt > chr5_teosinte_decrease_replace.txt	$ sed 's/?/-/g' chr6_teosinte_decrease.txt > chr6_teosinte_decrease_replace.txt	$ sed 's/?/-/g' chr7_teosinte_decrease.txt > chr7_teosinte_decrease_replace.txt	$ sed 's/?/-/g' chr8_teosinte_decrease.txt > chr8_teosinte_decrease_replace.txt	$ sed 's/?/-/g' chr9_teosinte_decrease.txt > chr9_teosinte_decrease_replace.txt	$ sed 's/?/-/g' chr10_teosinte_decrease.txt > chr10_teosinte_decrease_replace.txt**Separate out SNPs with unknown or multiple positions:**##	$ grep -E "multiple" joined_file.txt > multiple_maize.txt	$ grep -E "multiple" joined_teosinte.txt > multiple_teosinte.txt	$ grep -E "unknown" joined_file.txt > unknown_maize.txt	$ grep -E "unknown" joined_teosinte.txt > unknown_teosinte.txt**Manually remove any "multiple" SNP positions where chromosome was known**##	$ sed '/multiple/d' chr9_teosinte_decrease_replace.txt > chr9_teosinte_increase_replace_rm.txtModify for each file as needed